File: 96.py
class Solution:
    # @param {integer} n
    # @return {integer}
    def numTrees(self, n):
        count_mat = [[0 for x in range(n+2)] for y in range(n+2) ]
        for i in range(1,n+2):
            count_mat[i][i] = 1
            count_mat[i][i-1] = 1
        for s in range(1,n):
            for i in range(1,n-s+1):
                j = i + s
                temp_count = 0
                for k in range(i,j+1):
                    temp_count += count_mat[i][k-1] * count_mat[k+1][j]
                count_mat[i][j] = temp_count
        return count_mat[1][n]

        




sol = Solution()
print sol.numTrees(2)

File: 96.cpp
#include <iostream>
#include <cstring> 
using namespace std;

class Solution {
public:
    int numTrees(int n) {
        int A[n+2][n+2];
        memset(A,0,sizeof(A));
        for(int j =1; j<=n;j++){
            A[j+1][j] = 1;
            A[j][j] = 1;
        }
        for(int i =1;i<=n;i++){
            A[i][i-1] = 1;
        }
        
        // for(int i =0;i<=n+1;i++){
        //     for(int j =0;j<=n+1;j++){
        //         cout<<i<<' '<<j<<' '<<A[i][j]<<endl;
        //     }
        // }

        int j = 0;
        for(int k=1;k<=n-1;k++)
        {
            for(int i=1;i<=n-k;i++)
            {
                j = i+k;
                for(int s = i; s<=j;s++)
                {
                    A[i][j] += A[i][s-1]*A[s+1][j];
                    cout<<s<<' '<<i<<' '<<j<<' '<<A[i][j]<<endl;
                }
                cout<<i<<' '<<j<<' '<<A[i][j]<<endl;
            }
        }
        
        return A[1][n];
        
    }

    // new solution

    int numTrees(int n) {
        int store[n+1];
        memset(store,0,sizeof(store));
        store[0] = 1;
        store[1] = 1;
        
        for(int i =2; i<=n;i++)
        {
            for(int j = 1; j<=i;j++)
            {
                store[i] += store[j-1]*store[i-j];
            }
            
        }
        
        return store[n];
        
    }

};

int main()
{
    Solution sol;
    sol.numTrees(3);
    return 0;
}